react-task-1/src/components/Cards/CardsContainer/CardsContainer.js

```diff
@@ -1,32 +1,150 @@
-import React, { Component } from 'react';
+import React, { useState, useEffect } from 'react';
 import { apiCall } from '../../../api/mockedApi.js';
-import { Card } from '../Card/index.js'
+import { Card } from '../Card/index.js';
+import { CardsCreationForm } from '../CardsCreationForm/index.js';
+
 import styles from './CardsContainer.module.scss';
 
-export class CardsContainer extends Component {
+const initialState = {
+  title: {
+    name: 'title',
+    value: '',
+    type: 'text',
+    placeholder: 'Title',
+    hasError: false,
+    errorMessage: 'Title is required',
+  },
+  type: {
+    name: 'type',
+    value: '',
+    type: 'text',
+    placeholder: 'Type',
+    hasError: false,
+    errorMessage: 'Type is required',
+  },
+  imageUrl: {
+    name: 'imageUrl',
+    value: '',
+    type: 'text',
+    placeholder: 'Image Url',
+    hasError: false,
+    errorMessage: 'Url is required',
+  },
+  price: {
+    name: 'price',
+    value: '',
+    type: 'number',
+    placeholder: 'Price',
+    hasError: false,
+    errorMessage: 'Price is required',
+  },
+  country: {
+    name: 'country',
+    value: '',
+    type: 'text',
+    placeholder: 'Country',
+    hasError: false,
+    errorMessage: 'Country is required',
+  },
+}
 
-  state = {
-    cards: []
-  }
+export function CardsContainer() {
 
-  componentDidMount() {
+  const [state, setState] = useState({ cards: [] });
+  const [newCard, setNewCard] = useState(initialState);
+  
+  useEffect(() => {
     apiCall().then((data) => {
-      this.setState({ cards: data });
+      setState({ cards: data });
     });
+  }, []);
+
+  function deleteCard(id) {
+    const filtered = state.cards.filter(el => el.id !== id);
+    setState({ cards: filtered });
   }
 
-  render() {
-    const { cards } = this.state;
+  function newCardValidation () {
+    const reduced = Object.entries(newCard).reduce((acc, cur) => {
+      if (!cur[1].value) {
+        acc[cur[0]] = {
+          ...cur[1],
+          hasError: true,
+        }
+      } else {
+        acc[cur[0]] = {
+          ...cur[1],
+        }
+      }
+      return acc;
+    },{});
 
-    if (!cards.length) {
-      return (
-        <div className={styles.noCardsNotification}>No cards yet</div>
-      );
+    setNewCard(reduced);
+  }
+
+  function handleSubmit(event) {
+
+    event.preventDefault();
+
+    newCardValidation();
+
+    const children = event.target.children;
+
+    for (let i = 1; i < children.length - 1; i++) {
+      if (!children[i].value) {
+        children[i].focus();
+        return
+      }
     }
-    return (
-      <div className={styles.container}>
-        { cards.map((card) => <Card key={card.id} cardData={card} />)}
-      </div>
-    );
+    
+    const transformedNewCard = Object.entries(newCard).reduce((acc, cur) => {
+      acc[cur[0]] = cur[1].value;
+      acc['id'] = state.cards.length + 1;
+      return acc;
+    },{});
+
+    setState({ cards: [...state.cards, transformedNewCard] });
```

Comment:
```suggestion
    setState({
      cards: [
        ...state.cards,
      {
        title: newCard.title.value,
        type: newCard.type.value,
        imageUrl: newCard.imageUrl.value,
        price: newCard.price.value,
        country: newCard.country.value,
        id: state.cards.length + 1,
      }
    ]});
```
если сделать таким образом, то на 100-й строке не нужен будет цикл, что думаешь?

Reply:
согласен